### PR TITLE
Add DG components to the build system

### PR DESCRIPTION
**Pull request by timspainNERSC**
# Add DG components to the build system
## Fixes \#595
---
# Change Description

This PR moves the definition of the number of DG components and the CG degree from hard coded values spread between `ModelArrayDetails.cpp` and the dynamics module and kernel classes. These values are now defined as part of the CMake configuration step, with the CMake command line argument `-DDynamicsType=` with valid parameter values `DG1` and `DG2`. A value of `DG1` sets the advection DG degree to 1 and the number of advection DG components to 3. A value of `DG2` sets the advection DG degree to 2 and the number of advection DG components to 6. In both cases the CG degree is set to 2 and the number of DG stress advection components to 8.

These values are implemented as values for preprocessor macros that are applied to all of the source files compiled into the `nextsimlib` compilation target. This is then used by the model binary and most of the unit tests.

The `ModelArrayDetails.cpp` file defines the number of components of the DG-related `ModelArray`s based on these values, but also defines values equivalent to the `DG1` option as defaults.

In the dynamics, the number of advection DG components is used as the template parameter in instantiating the relevant dynamics kernel. The CG degree is used in the `CGDynamicsKernel` class to as the `CGdegree` template parameter and to calculate the number of DG components for stress. This latter value is calculated using the `CG2DGCOMPONENTS` function. These two numbers of DG components are used as the template parameters for the `DynamicsKernel` class. Within these various dynamics kernel classes the template parameters for the underlying dynamics classes such as `DGVector` and `CGVector` are set.

---
# Test Description

Tests still compile and run.

---
# Documentation Impact

Information on the options added to installation.rst.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -90,14 +90,54 @@ set(ScriptDirectory "${PROJECT_SOURCE_DIR}/scripts")
...
+# Dynamics type. Defaults to DG1.
+# Available options are:
+# DG1 Discontinuous Galerkin, degree 1 (3 components)
+# DG2 Discontinuous Galerkin, degree 2 (6 components)
+set(isDG FALSE)
+
+# Replace a missing dynamics type with DG1
+if (NOT DynamicsType)
+    message("Defaulting dynamics to DG1")
+    set(DynamicsType "DG1")
```
Comment (einola) [inline CMakeLists.txt:102]:
```suggestion
    set(DynamicsType "DG2")
```

I think we'll most often run the model like this, but I'll admit that's not clear yet.

**Review by einola — APPROVED** [review]
Two small comments, but looks fine otherwise.

File: CMakeLists.txt [hunk 2]
```diff
@@ -90,14 +90,54 @@ set(ScriptDirectory "${PROJECT_SOURCE_DIR}/scripts")
...
+endif()
+
+message("Dynamics is set to ${DynamicsType}")
+
+if (DynamicsType STREQUAL "DG1")
+    set(isDG TRUE)
+    set(DGComp 3)
+    set(CGDegree 2)
```
[the remark this answers is not in the document]
Comment (timspainNERSC) [reply CMakeLists.txt:110]: Oh, maybe I am misunderstanding something. I will think further.

File: CMakeLists.txt (hunk 1 above)
Comment (timspainNERSC) [reply CMakeLists.txt:102]: Agreed on both counts.

File: CMakeLists.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (timspainNERSC) [reply CMakeLists.txt:110]: Yes, I think the suggested change is correct.

**Comment by timspainNERSC** [comment]
Subsumed into #606

**Comment by timspainNERSC** [comment]
Subsumed into #606

**Pull request closed without merging**